frameworks/keyed/mikado/src/main.js

```diff
@@ -0,0 +1,38 @@
+import Mikado from "../node_modules/mikado/src/mikado.js";
+import { route } from "../node_modules/mikado/src/event.js";
+import tpl_app from "./template/app.js";
+import tpl_item from "./template/item.js";
+import buildData from "./data.js";
+
+Mikado.once(document.body, tpl_app).eventCache = true;
+
+// This implementation is using a full declarative paradigm,
+// where updates are delegated by same indizes as used for the data access.
+
+let data;
+const state = {};
+const view = new Mikado(tpl_item, { mount: document.getElementById("tbody"), state });
+const event = { stop: true };
+
+route("run", () => view.render(data = buildData(1000)), event);
+route("runlots", () => view.render(buildData(10000)), event);
+route("add", () => view.append(data = buildData(1000)), event);
+route("update", () => {
+    for(let i = 0; i < data.length; i += 10){
+        data[i].label += " !!!";
+        view.update(i, data[i]);
+    }
+}, event);
+route("clear", () => view.clear(), event);
+route("swaprows", () => {
+    const tmp = data[1];
+    view.replace(1, data[1] = data[998]);
+    view.replace(998, data[998] = tmp);
```

Comment:
@krausest @ts-thomas 

should we have an additional flag for implementations like this?

i would not consider this type of implementation to be declarative, and is not too far from jQuery where you update the data and the view together to keep them in sync :)

Reply:
Yes, that's pretty close to the metal.
We could use #772 for that or invent a new flag, maybe something like "Implementation uses imperative code for reconciliation".
Your opinions?

Reply:
i think either way is fine. changing #772 to say "imperative UI updates" feels better, just to avoid additional flags. but i suspect authors may want to distinguish between the strategies.

Reply:
Let's reduce to what it is. It initialize an update by a numeric index. Therefore every library which uses a numeric index to initialize this update task needs to be flagged also. Then I'm fine with it. 

For example from "million":

```js
const swapRows = () => {
  if (list.length > 998) {
    const item = list[1];
    list[1] = list[998];
    list[998] = item;
  }
  update();
  return false;
};
```

Is also initializing this update by index. The only difference is when it's done, and this is a specification of the library.

Reply:
this test has always been about testing a general reconciliation strategy (because that's what react does poorly). otherwise, the literal "swap" by itself is a pointless/stupid test. we can probably agree that it's stupid if it doesnt test anything meaningful?

we should just replace this with a seeded random "shuffle 100 data items into new positions", so that implementations which treat it as a "swap" test look obviously absurd.

Reply:
I will replace it and make a pull request tomorrow.

I would like to explain why it isn't very useful when expecting a specific (almost data-driven) style.

Let's take the example from million. I don't know how the library is implemeneted, but I have a draft of a "transaction" feature for Mikado reactive proxy.

A look behind the scene is almost like:

```js
// a task queue under the hood
const task = [];

// a proxified list
const item = list[1];
// the proxy trap do: task.push(update)
list[1] = list[998];
// the proxy trap do: task.push(update)
list[998] = item;

// execute the update by:
update();
// will run under the hood:
task.forEach(update => update())
```

Unrolled it just do:

```js
update(task)
update(task)
```

wich is technically the same as:

```js
update(index)
update(index)
```

The only difference is the specification of the library which just uses two different approaches, sharing the same complexity.

Reply:
> this test has always been about testing a general reconciliation strategy (because that's what react does poorly). otherwise, the literal "swap" by itself is a pointless/stupid test. we can probably agree that it's stupid if it doesnt test anything meaningful?
> 
> we should just replace this with a seeded random "shuffle 100 data items into new positions", so that implementations which treat it as a "swap" test look obviously absurd.

It's not the implementation, the problem is the nature of the test. A better concept is what I'm using in my benchmark, where the problem isn't known, because data is coming from an external authority. So the whole test implementation code reduces to something like:

```js
suite["mikado"] = function(items){
    mikado.render(items);
};
```

Anything else is useless for me as a benchmark scenario.

Reply:
> we should just replace this with a seeded random "shuffle 100 data items into new positions", so that implementations which treat it as a "swap" test look obviously absurd.

Why hasn't this been used? Is there a discussion on it?